cou/steps/plan.py

```diff
@@ -375,6 +379,55 @@ async def generate_plan(analysis_result: Analysis, args: CLIargs) -> UpgradePlan
     return plan
 
 
+async def filter_hypervisors_machines(args: CLIargs, analysis_result: Analysis) -> list[Machine]:
+    """Filter the hypervisors to generate plan and upgrade.
+
+    :param args: CLI arguments
+    :type args: CLIargs
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :return: hypervisors filtered to generate plan and upgrade.
+    :rtype: list[Machine]
+    """
+    hypervisors_machines = await _get_upgradable_hypervisors_machines(args.force, analysis_result)
+
+    if cli_machines := args.machines:
+        return [machine for machine in hypervisors_machines if machine.machine_id in cli_machines]
+
+    if cli_hostnames := args.hostnames:
+        return [machine for machine in hypervisors_machines if machine.hostname in cli_hostnames]
+
+    if cli_azs := args.availability_zones:
+        return [machine for machine in hypervisors_machines if machine.az in cli_azs]
```

Comment:
non-blocking: though this function is already pretty readable, it has similar code across these conditions which seems redundant. An idea for refactoring:
```python
cli_filters = [
        ('machine_id', args.machines),
        ('hostname', args.hostnames),
        ('az', args.availability_zones),
    ]
for attr, filter in cli_filters:
    if filter:
          return [machine for machine in hypervisors_machines if getattr(machine, attr) in filter]
```
This is also easily extendable if we need to add more filters in the future

Reply:
It's extendable only for machines filters, I do not think we will have much, and to me, it's less readable with more space for an issue.

Reply:
IMO this is less prone to error because you don't need to write this list comprehension multiple times `[machine for machine in hypervisors_machines if machine.az in cli_azs]` and have the chance to mistype something. It's also easier for reviews because they are using a unified way of iteration.